api/spec/requests/spree/api/line_items_spec.rb

```diff
@@ -95,6 +95,12 @@ module Spree::Api
         expect(response.status).to eq(201)
       end
 
+      it '#create calls #invalid_resource! if adding a line item fails validation' do
```

Comment:
Instead of testing the implementation, I think, in this case, we can test the actual behavior. I mean sending invalid data and checking that the response contains an error. Maybe we could try sending `-1` as quantity?

Reply:
I thought of this as well, but I am afraid, we can't do it through variant or quantity. LineItem's quantity gets normalized to zero upon validation if it's `nil` or negative:
https://github.com/solidusio/solidus/blob/6dae6359e5d8cf7a83aa473e59f7888d2090f6f1/core/app/models/spree/line_item.rb#L25
https://github.com/solidusio/solidus/blob/6dae6359e5d8cf7a83aa473e59f7888d2090f6f1/core/app/models/spree/line_item.rb#L137-L139

So in away, this is really a non-occurring problem, but it's problematic if validations on `LineItem` changed.

Reply:
Oh, I see. It looks like there's no point in testing the unhappy path, as that's something not considered by our business logic (which explains why we never found an issue here). I wonder if it makes sense to make the controller code less defensive and acknowledge that it should never fail.

Reply:
It makes partial sense to simplify the endpoint if cannot currently fail, but then it's always relying on that fact, and therefore if `LineItem` received a new validation, this endpoint would start failing and it would not be realised by the specs 🤔 

Also about perhaps extracting the following line to a validation on `LineItem`? If we did this, we would need this this API endpoint to handle LineItem errors:
https://github.com/solidusio/solidus_starter_frontend/blob/fa62873bf7d06ef8054de1dbba8ca5bee27a1cfc/templates/app/controllers/cart_line_items_controller.rb#L19

Reply:
I wouldn't be defensive trying to anticipate what can happen in the future. Our system should reflect the current state of things. However, we can also argue if it makes sense to normalize the quantity to zero when a negative number is given. I think a consumer would be better served if an explicit error is given in that case. Happy to hear other thoghts. cc @solidusio/core-team

Reply:
I tried to change that part in the past, without a respectable patch to be proposed. I think that normalization shouldn't happen in a before_validation. It would be great to move that code in a validation and I think that this comes from the feature that allows to remove line items if the quantity is 0, which is more of a UX problem and should not be a model responsibility. 

Solving this problem will also clean up several tangential parts of the codebase, like these specs, but of course, it's not part of this PR's purpose. 


That said, for this specific spec, what if we add some fake errors to the line item and check the error is present in the response?

```ruby
allow_any_instance_of(Spree::LineItem).to receive(:errors).and_return("something")
post spree.api_order_line_items_path(order), params: { line_item: { variant_id: product.master.to_param, quantity: 1 } }
expect(json_response["error"]).to eq('something')
```

Of course, this is just pseudo-code I didn't actually check, only useful to explain better what I meant. 😆

Reply:
I am happy to open up a PR to investigate removing the quantity normalization if desired by the core team. If you want, we could keep this PR open until the change is actually required (such as removing the quantity normalization).

Reply:
If this PR solves a real store problem I won't block merging it. 

It would be great to see if we can get rid of the normalization and just use some plain validations; if you want to have a quick conversation around that before starting, please ping me privately, I would be happy to help.

Reply:
Yeah, I agree. We can merge this one and work on the actual fix afterward.

Reply:
@kennyadsl solution is a good compromise for now

Reply:
I will update the spec to use a quantity of 0 once we update validation.